### PR TITLE
nixos: make it easy to apply kernel patches

### PR DESCRIPTION
**Pull request by cstrahan**
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This makes it easy to specify kernel patches:

```
boot.kernelPatches = [ pkgs.kernelPatches.ubuntu_fan_4_4 ];
```

To make the `boot.kernelPatches` option possible, this also makes it easy
override the kernel and/or packages within a linuxPackages set. For
example:

```
pkgs.linuxPackages.extend (self: super: {
  kernel = super.kernel.override {
    kernelPatches = super.kernel.kernelPatches ++ [ pkgs.kernelPatches.ubuntu_fan_4_4 ];
  };
});
```

For additional context, you can read my letter to the mailing list: http://lists.science.uu.nl/pipermail/nix-dev/2016-May/020342.html

---

/cc @thoughtpolice @nshalman

### CODE REVIEW

**Comment by magnetophon** [comment]
@henrytill maybe good for [musnix](https://github.com/musnix/musnix)?

**Comment by cstrahan** [comment]
I'm tempted to factor out the recursive bits into something reusable.

Would something like this in `lib/trivial.nix` make sense? (aside from naming, for which I'd like suggestions)


```nix
  # Compose two extension functions:
  #
  #     nix-repl> h = self: super: { bar = "<${super.bar}>"; }
  #
  #     nix-repl> fix (extends (composeExtensions h g) f)
  #     { bar = "<bar>"; foo = "foo + "; foobar = "foo + <bar>"; }
  composeExtensions = f: g: self: super:
    let attrs = g self super;
        super' = super // attrs;
    in attrs // f self super';

  # Create an overridable, recursive attribute set. For example:
  #
  #     nix-repl> obj = lib.makeObject (self: { })
  #
  #     nix-repl> obj
  #     { override = «lambda»; }
  #
  #     nix-repl> obj = obj.override (self: super: { foo = "foo"; })
  #
  #     nix-repl> obj
  #     { foo = "foo"; override = «lambda»; }
  #
  #     nix-repl> obj = obj.override (self: super: { foo = super.foo + " + "; bar = "bar"; foobar =
  #     self.foo + self.bar; })
  #
  #     nix-repl> obj
  #     { bar = "bar"; foo = "foo + "; foobar = "foo + bar"; override = «lambda»; }
  makeObject = f:
    let
      construct = overrides: self:
        let
          override = f:
            let overrides' = composeExtensions f overrides;
            in fix (extends overrides' (construct overrides'));
        in f self // { inherit override; };
    in fix (construct (self: super: {}));
```

**Comment by cstrahan** [comment]
**I pushed up a second commit that shows what this would look like with the recursive bits factor out.**

What do you think? I'd like some feedback before this gets merged.

**Comment by cstrahan** [comment]
Actually, I suppose this would be a better/leaner implementation of the the `makeObject` function that I proposed above:

```nix
  makeObject = rattrs:
    let
      construct = rattrs:
        fix (self: rattrs self // {
          override = f: construct (extends f rattrs);
        });
    in construct rattrs;
```

**Comment by cstrahan** [comment]
@dezgeg What are your thoughts on the `makeObject` function?

**Comment by joachifm** [comment]
Does
```nix
{
  boot.kernelPackages = with pkgs;
    let self = linuxPackagesFor (kernel.override { kernelPatches = [ /* ... */ ]; }) self; in self;
}
```
not accomplish the same thing?

**Comment by cstrahan** [comment]
@joachifm 

That can definitely work, however:

1. That doesn't compose well. For example, you can't have the grsecurity kernel module set that option while simultaneously setting that option yourself to apply additional patches -- you can only have one or the other.
2. That's a whole lot of work just to apply some patches, and it requires that you know about `linuxPackagesFor` and how it works (and the oddity of having to pass `self` yourself).

As an example, when we want to extend/override Haskell packages, we aren't required to re-implement `haskell` via `haskell = pkgs.callPackage <nixpkgs/pkgs/haskell-packages.nix> { /* ... overrides ... */ };`. Instead, we just provide an "overrides" function, thus keeping knowledge of the implementation out of our configurations. Presently, we can't extend the `linuxPackages` -- to add or modify a package in that set, we have to completely re-evaluate the expressions that make up the set.

I believe this PR makes things compose better and makes the kernel packages extensible without leaking knowledge of the implementation details into your configuration.

**Comment by joachifm** [comment]
This interface is better, no doubt about it. I'm not convinced composing kernel patches like this is really sane, but it is quite user unfriendly to do it at all right now, so ...

File: lib/trivial.nix [hunk 1]
```diff
@@ -53,6 +53,41 @@ rec {
...
+  #     nix-repl> obj
+  #     { bar = "bar"; foo = "foo + "; foobar = "foo + bar"; override = «lambda»; }
+  makeObject = rattrs:
+    let
+      construct = rattrs:
+        fix (self: rattrs self // {
+          override = f: construct (extends f rattrs);
+        });
+    in construct rattrs;
+
```
Comment (Ericson2314) [inline lib/trivial.nix]: Off topic: but we should make a `fix.nix`/`combinator.nix`/etc for all these. It's getting rather presumptuous to have all this is `trivial.nix`. :)

File: nixos/modules/system/boot/kernel.nix [hunk 2]
```diff
@@ -4,6 +4,8 @@ with lib;
 
 let
 
+  kernelPatches = config.boot.kernelPatches;
```
Comment (Ericson2314) [inline nixos/modules/system/boot/kernel.nix]: I'm a big fan of `let inherit`, as in `inherit (config.boot) kernelPatches;`

**Comment by cstrahan** [comment]
> I'm not convinced composing kernel patches like this is really sane, but it is quite user unfriendly to do it at all right now, so ...

Right - the more patches you throw in, the greater the chance that one of the patches won't apply (or worse, will apply, but will be buggy). For the simple, 99.9% case where someone just wants to add their patches to the default `linuxPackages` value, the user won't run into any problems, and this provides a much lower barrier to getting their patches applied.

And, of course, having a way to add packages to the set of `linuxPackages` is only a good thing, given that currently `linuxPackages` is a closed set at the moment.

File: lib/trivial.nix (hunk 1 above)
Comment (cstrahan) [inline lib/trivial.nix]: Yeah, it's not looking so trivial anymore :). I can move things around, if we so wish.

File: nixos/modules/system/boot/kernel.nix (hunk 2 above)
Comment (cstrahan) [inline nixos/modules/system/boot/kernel.nix]: Thanks, just pushed that change.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -10805,26 +10807,25 @@ in
...
+  linuxPackages_3_10_tuxonice = self.linuxPackagesFor self.linux_3_10_tuxonice;
+  linuxPackages_3_12 = recurseIntoAttrs (self.linuxPackagesFor self.linux_3_12);
+  linuxPackages_3_14 = recurseIntoAttrs (self.linuxPackagesFor self.linux_3_14);
+  linuxPackages_3_18 = recurseIntoAttrs (self.linuxPackagesFor self.linux_3_18);
+  linuxPackages_4_1 = recurseIntoAttrs (self.linuxPackagesFor self.linux_4_1);
+  linuxPackages_4_3 = recurseIntoAttrs (self.linuxPackagesFor self.linux_4_3);
+  linuxPackages_4_4 = recurseIntoAttrs (self.linuxPackagesFor self.linux_4_4);
+  linuxPackages_4_5 = recurseIntoAttrs (self.linuxPackagesFor self.linux_4_5);
+  linuxPackages_testing = recurseIntoAttrs (self.linuxPackagesFor self.linux_testing);
   linuxPackages_custom = {version, src, configfile}:
```
Comment (Ericson2314) [inline pkgs/top-level/all-packages.nix]: Damn that's some crazy indenting. IIUC our style these days + removing unneeded bindings would make it more like:
```nix
let # the `let` is just for syntax highlighting
  linuxPackages_custom = {version, src, configfile}:
    recurseIntoAttrs (self.linuxPackagesFor (self.linuxManualConfig {
      inherit version src configfile;
      allowImportFromDerivation = true;
    }));
```

Comment (cstrahan) [inline pkgs/top-level/all-packages.nix]: I wanted to keep the diff easy to view, but I suppose I could also reformat things. I think I'd prefer to do that in a separate PR, though.

Comment (Ericson2314) [inline pkgs/top-level/all-packages.nix]: Yeah there's "we're touching the lines anyways" on one and and "PR mission creep on the other.

File: lib/trivial.nix [hunk 4]
```diff
@@ -53,6 +53,41 @@ rec {
   # argument, but it's nice this way if several uses of `extends` are cascaded.
   extends = f: rattrs: self: let super = rattrs self; in super // f self super;
 
+  # Compose two extension functions:
+  #
+  #     nix-repl> h = self: super: { bar = "<${super.bar}>"; }
+  #
+  #     nix-repl> fix (extends (composeExtensions h g) f)
+  #     { bar = "<bar>"; foo = "foo + "; foobar = "foo + <bar>"; }
+  composeExtensions = f: g: self: super:
```
Comment (Ericson2314) [inline lib/trivial.nix]: Is this the same as `f: g: self: (extend f (extend g self)`?

**Comment by Ericson2314** [comment]
Big +1 from me! I think in a follow-up PR it should be fairly straightforward to deal with with the multiple `linuxPackages` too----`platform` needs an overhaul anyways (#10874). But that is out of scope for this PR.

File: lib/trivial.nix (hunk 4 above)
Comment (cstrahan) [inline lib/trivial.nix]: No, that doesn't quite do what we want it to:

```
(extend f (extend g self)
#                   ^
#------------------/
```

That needs to be a function taking it's fixpoint (e.g. the `rattrs` naming convention), as that's what `extends` expects.

**Comment by cstrahan** [comment]
@Ericson2314 Thanks for reviewing :)

File: lib/trivial.nix (hunk 4 above)
Comment (Ericson2314) [inline lib/trivial.nix]: Ah yes. That said, it appears you no longer use this function? `f: g: rattrs: (extend f (extend g rattrs)` seems to be adequate for the situations where I imagine this would be used, even if it is not exactly the same.

Comment (cstrahan) [inline lib/trivial.nix]: Yeah, I think I'll remove it for now.

Comment (Ericson2314) [inline lib/trivial.nix]: Sounds good.

**Comment by cstrahan** [comment]
I see a potential problem regarding the above definitions of `makeObject`: its idea of `override` is different than `callPackage`'s.

`callPackage` provides an `override` attribute with allows you to change the *inputs* to a function, while `makeObject` provides an `override` attribute that allows you to modify and/or extend the *outputs*. If you compose `callPackage` over `makeObject`, the `override` function from `callPackage` clobbers the one from `makeObject`. Even with that problem aside, the conflation of the word `override` is potentially confusing.

To address that last point, I'm thinking that `makeObject` should instead provide an `extend` attribute, that way the word `override` isn't overloaded.

That also happens to resolve the first point, but that does make me wonder about how combinations of `override` and `extend` should behave. With that change alone, doing `(somePkgSet.extend (sef: super: { /* ... */ }).override { /* ... args ... */ }` will result in the extensions from `extFn` being ignored, as `override` will result in the package function being re-evaluated. In a similar vein, the result of `extend` will lose the `override` attribute. Adding a new function analogous to `callPackage` but for recursive, extensible package sets would fix that. I see two alternative implementations:

---

**Option 1 - don't preserve extensions when overriding**

```
callRecPackagesWith = autoArgs: fn: args:
  let
    f = if builtins.isFunction fn then fn else import fn;
    auto = builtins.intersectAttrs (builtins.functionArgs f) autoArgs;
    go = args: lib.makeObject (self:
      f (auto // args) self // { override = newArgs: go (args // newArgs); }
    );
  in go (auto // args);

# Examples (common between both implementations):
fn = { foo, bar, baz, quux }: self: { inherit foo bar baz quux; };
crp = callRecPackagesWith { foo = "foo"; bar = "bar"; baz = "baz"; quux = "quux"; };
a = crp fn { foo = "AAA"; };
b = a.override { bar = "BBB"; };
c = b.extend (self: super: { baz = super.baz + " CCC"; });
d = c.override { baz = "DDD"; };
```

```
nix-repl> :l ./lib
Added 321 variables.

nix-repl> a
{ bar = "bar"; baz = "baz"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> b
{ bar = "BBB"; baz = "baz"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> c
{ bar = "BBB"; baz = "baz CCC"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> d
{ bar = "BBB"; baz = "DDD"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }
```

Note how in `d` we lose the extension on `baz`.

---

**Option 2 - do preserve extensions when overriding**

```
callRecPackagesWith = autoArgs: fn: args:
  let
    f = if builtins.isFunction fn then fn else import fn;
    auto = builtins.intersectAttrs (builtins.functionArgs f) autoArgs;
    go = extensions: args:
      lib.fix (self:
        lib.extends extensions (f (auto // args)) self
          // { override = newArgs: go extensions (args // newArgs); }
          // { extend = extFn: go (lib.composeExtensions extFn extensions) args; }
    );
  in go (self: super: {}) (auto // args);

# Examples (common between both implementations):
fn = { foo, bar, baz, quux }: self: { inherit foo bar baz quux; };
crp = callRecPackagesWith { foo = "foo"; bar = "bar"; baz = "baz"; quux = "quux"; };
a = crp fn { foo = "AAA"; };
b = a.override { bar = "BBB"; };
c = b.extend (self: super: { baz = super.baz + " CCC"; });
d = c.override { baz = "DDD"; };
```


```
nix-repl> :l ./lib
Added 321 variables.

nix-repl> a
{ bar = "bar"; baz = "baz"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> b
{ bar = "BBB"; baz = "baz"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> c
{ bar = "BBB"; baz = "baz CCC"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> d
{ bar = "BBB"; baz = "DDD CCC"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }1
```

Note how in `d` we *keep* the extension on `baz`, and wee see that `d.baz == "DDD CCC"`.

---

**TL;DR**

I think I'd like to have the second `callRecPackagesWith` (though I'd like a less atrocious name for it), and I'd like to rename `makeObject`'s `override` to `extend`.

@Ericson2314 How does that sound?

**Comment by cstrahan** [comment]
Note that this PR doesn't necessarily need a solution to the `callPackage` scenario, as the `linuxPackages` attrset doesn't use `callPackage`. However, I figured if I was going to introduce a new function like `makeObject`, I might as well get the design right upfront so we can start to use it in more places.

**Comment by cstrahan** [comment]
Another observation while I'm at it, though this is slightly off topic and not holding this PR back:

At a later time, we could transition `mkDerivation` to take a function that receives its fixpoint. This would allow us to define a similar `extend` on derivations, where `extend` would give us something like `overrideDerivation`, only it could behave a lot better, while still allowing for chaining `override`.

We could slowly transition by having `mkDerivation` use `builtins.isFunction` to discern wether it's just an attrset or a function expecting its fixpoint.

I think this reaffirms my decision on the naming, `override` vs `extend`.

**Comment by Ericson2314** [comment]
@cstrahan Nice write up! To recap what I said on IRC, I think the second option is better on the merits that `.override` and `.extend` represent mock-imperative mock-mutating idioms, so "state" should be preserved.

I think the second option could also be rewritten to look like this. I believe the single occurence of `auto //` will make it more efficient, and I removed `composeExtensions` because I still find it a bit extraneous.
```nix
let # for colors
callRecPackagesWith = autoArgs: fn: origArgs:
  let
    f = if builtins.isFunction fn then fn else import fn;
    auto = builtins.intersectAttrs (builtins.functionArgs f) autoArgs;
    go = rattrs: args:
      lib.fix (self:
        rattrs ((f args) self)
          // { override = newArgs: go extensions (args // newArgs); }
          // { extend = newRattrs: go (lib.extend newRattrs rattrs) args; }
    );
  in go (self: {}) (auto // origArgs);
```

One remaining question is, should each `self: super: {}` overrrider also take additional arguments like the inital call package, so we have some sort of callpackaging reader monad at each override? That probably makes no sense so I'll try to write it :).

**Comment by cstrahan** [comment]
@Ericson2314 

Ah, I don't know why my brain keeps going down the `lib.composeExtensions` route -- I somehow find that easier to reason about, until I work through the alternative -- and then former looks complicated and the latter looks simple :).

Anyway, I had to tweak your code just a bit, but yes, I think that's better:

```nix
  let # for colors
  callRecPackagesWith = autoArgs: fn: origArgs:
    let
      f = if builtins.isFunction fn then fn else import fn;
      auto = builtins.intersectAttrs (builtins.functionArgs f) autoArgs;
      go = rattrs: args:
        lib.fix (self:
          rattrs ((f args) self)
            // { override = newArgs: go rattrs (args // newArgs); }
            // { extend = newRattrs: go (lib.extends newRattrs rattrs) args; }
      );
    in go (self: self) (auto // origArgs);
```

> One remaining question is, should each self: super: {} overrrider also take additional arguments like the inital call package, so we have some sort of callpackaging reader monad at each override? That probably makes no sense so I'll try to write it :).

Hehe, yeah, I might need to see an example :).

**Comment by cstrahan** [comment]
@Ericson2314 Oh, if you mean automatically adding or providing `callPackage` scoped to `self`, I think it's sufficient to let the caller do `{ callPackage = callPackageWith self; /* ... other attrs ... */ }` -- that's what I do in this PR for `linuxPackages`.

**Comment by Ericson2314** [comment]
Well, perhaps it is easier to first envision with `lib.composeExtensions` --- I've had the enormous benefit of the second draft however :).

Ok here it is:

Also, I think it's easier to just use `{ foo, bar, ... }: ...` than this intersection bussiness.
```nix
let # for colors
# autoArgs, origArgs, args : { b }
# fn : { b } -> self @ { a } -> { a } 
# newReaderRattrs : { b } -> self @ { a } -> super @ { a } -> { a }
# returns : { a }
callRecPackagesWith = autoArgs: fn: origArgs:
  let
    f = if builtins.isFunction fn then fn else import fn;
    go = readerRattrs: args:
      lib.fix (self:
        readerRattrs args self
          // { override = newArgs: go readerRattrs (args // newArgs); }
          // { extend = newReaderRattrs: go
                 (params: lib.extend (newReaderRattrs params) (rattrs params))
                 args;
             }
    );
  in go f (autoArgs // origArgs);
```
Note that now `autoArgs` and `origArgs` are treated exactly the same.

**Comment by Ericson2314** [comment]
For reference, with the intersection stuff its
```nix
let # for colors
# autoArgs, origArgs, args : { b }
# fn : { b } -> self @ { a } -> { a } 
# newReaderRattrs : { b } -> self @ { a } -> super @ { a } -> { a }
# returns : { a }
callRecPackagesWith = autoArgs: fn: origArgs:
  let
    f = if builtins.isFunction fn then fn else import fn;
    auto = f: builtins.intersectAttrs (builtins.functionArgs f) autoArgs;
    app = f: args: f (auto f autoArgs // args);
    go = readerRattrs: args:
      lib.fix (self:
        readerRattrs args self
          // { override = newArgs: go readerRattrs (args // newArgs); }
          // { extend = newReaderRattrs: go
                 (params: lib.extend (app newReaderRattrs params) (rattrs params))
                 args;
             }
    );
  in go (app f) (origArgs);
```
not as bad as I thought.

**Comment by cstrahan** [comment]
Interesting. I don't think you have recompute `auto` every time, and can do away with `app`. Here's my rendition:

```nix
  let # for colors
  callRecPackagesWith = autoArgs: fn: origArgs:
    let
      f = if builtins.isFunction fn then fn else import fn;
      auto = builtins.intersectAttrs (builtins.functionArgs f) autoArgs;
      go = readerRattrs: args: allArgs:
        lib.fix (self:
          readerRattrs (f args self)
            // { override = newArgs: go readerRattrs (args // newArgs) (allArgs // newArgs); }
            // { extend = extFn: go
                   (let auto = builtins.intersectAttrs (builtins.functionArgs extFn) allArgs;
                        auto' = if auto == {} then allArgs else auto;
                    in lib.extends (extFn auto') readerRattrs)
                   args 
                   allArgs;
               }
        );
    in go (self: self) (auto // origArgs) (autoArgs // origArgs);

  fn = { foo, bar, baz, quux }: self: { inherit foo bar baz quux; };
  crp = callRecPackagesWith { foo = "foo"; bar = "bar"; baz = "baz"; quux = "quux"; };
  a = crp fn { foo = "AAA"; };
  b = a.override { bar = "BBB"; };
  c = b.extend ({bar}: self: super: { baz = super.baz + " CCC"; quux = bar + " quux"; });
  d = c.override { baz = "DDD"; };
  e = d.extend (params: self: super: { fnord = "fnord " + params.baz; baz = "bazinga"; });
```

```
nix-repl> :l ./lib
Added 322 variables.

nix-repl> a
{ bar = "bar"; baz = "baz"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> b
{ bar = "BBB"; baz = "baz"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "quux"; }

nix-repl> c
{ bar = "BBB"; baz = "baz CCC"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "BBB quux"; }

nix-repl> d
{ bar = "BBB"; baz = "DDD CCC"; extend = «lambda»; foo = "AAA"; override = «lambda»; quux = "BBB quux"; }

nix-repl> e
{ bar = "BBB"; baz = "bazinga"; extend = «lambda»; fnord = "fnord DDD"; foo = "AAA"; override = «lambda»; quux = "BBB quux"; }
```

That's probably a little overkill, and probably pretty slow... but it's definitely interesting.

**Comment by Ericson2314** [comment]
Yeah I'm not sure where to draw the line on combinator rediculousness. One thing to consider is when the inputs of one package set are drawn from the ouputs of the the containing package set, one can always just use `extend` on the outer package set instead of override on the inner. In that case, plain `mkObject` with a `callPackage` that doesn't attach any `override` field might work best.

**Comment by edolstra** [comment]
I don't like the name `makeObject` very much, Nix not being an object-oriented language :-)

Apart from that, there is an kernel configuration example in `nixos/doc/manual/configuration/linux-kernel.xml` that may need to updated.

File: lib/trivial.nix [hunk 5]
```diff
@@ -53,6 +53,30 @@ rec {
...
+  #
+  #     nix-repl> obj
+  #     { bar = "bar"; foo = "foo + "; foobar = "foo + bar"; override = «lambda»; }
+  makeObject = rattrs:
+    let
+      construct = rattrs:
+        fix (self: rattrs self // {
+          override = f: construct (extends f rattrs);
+        });
+    in construct rattrs;
```
Comment (roconnor) [inline lib/trivial.nix]: Let's simplify your definition of makeObject

rattrs: let
      construct = rattrs:
        fix (self: rattrs self // {
          override = f: construct (extends f rattrs);
        });
    in construct rattrs;

construct doesn't depend the the parameter, so we can float the let block out of the function.

let construct = rattrs:
        fix (self: rattrs self // {
          override = f: construct (extends f rattrs);
        });
    in rattrs: construct rattrs;

We can eta-contract the (ratters: construct rattrs) expression:

let construct = rattrs:
        fix (self: rattrs self // {
          override = f: construct (extends f rattrs);
        });
    in construct

Maybe use extends:

let construct = rattrs:
        fix (extends {override = f: construct (extends f rattrs)} rattrs)
    in construct

**Comment by roconnor** [comment]
In my mind, I like thinking of the open recursive functions as "classes" and the fixpoint of such "classes" as "objects" (which are really attribute sets).  See <http://r6.ca/oop.nix>.  I slightly prefer the name `newObject` to `makeObject`, but I'm not going to bikeshed over the issue.

I wonder if it is safe to use the attribute name `override`?  If this is going to be general purpose than maybe we need to add more underscores around it, like `__override__`.

Lastly, I'm not sure what `makeObject` brings to the table versus the existing `fix'` function.  Can't we simply define a library function:

    lib.overrideObject = ext: obj: fix' (lib.extends ext obj.__unfix__)

to override fixpoints that are defiend by `fix'`?

**Comment by Ericson2314** [comment]
@roconnor it closes over `__unfix__` rather than storing it in the attrset, which seems cleaner to me. Therefore I'd lean more towards the opposite and removing `fix'`.

**Comment by roconnor** [comment]
@Ericson2314  I don't follow you. 

`makeObject` stores a value in the `override` attribute in the resulting attrset that is very similar to `__unfix__`.  The main difference is that `makeObject`'s `override` attribute is available at class definition time via `self.override`, but this seems more confusing than helpful.  The other difference is that `__unfix__` is slightly more powerful as `override` is definable from `__unfix__` but I don't think think `__unfix__` is definable from `override`.

**Comment by Ericson2314** [comment]
That's all correct. I've only seen `fix'`/`__unfix__` used to define an override field like that (including as opposed to using a function bound elsewhere like `overrideObject`), and for this use case `mkObject`  avoids the intermediate `__unfix__` binding by closing over the old function directly.

**Comment by cstrahan** [comment]
Re: eta-contraction and such -- yeah, I've been a little exhausted lately. Brain isn't firing on all cylinders :).

> The main difference is that `makeObject`'s `override` attribute is available at class definition time via `self.override`, but this seems more confusing than helpful.

I don't know if it's so much confusing as it is unhelpful (can't imagine someone will want to do `makeObject (self: /*...*/ self.override /*...*/)`).

> The other difference is that `__unfix__` is slightly more powerful as override is definable from `__unfix__` but I don't think think `__unfix__` is definable from override.

I think I agree with @Ericson2314's response to this: I don't think I've seen any use other than providing the whole `.override (self: super: { /*...*/ })` thing. Granted, I've already admitted to being a little groggy, so perhaps I've missed something that makes `__unfix__` in itself valuable. I guess the only case where `__unfix__` gets interesting is when used like `lib.fix (self: foo.__unfix__ (bar.__unfix__ self))`, though I can't really imagine a scenario where we want to compose two separate package sets in that way. It's quite possible that I'm not imaginative enough, or that `makeObject` would be rendered insufficient for some exotic non-package set purpose.

Thoughts, @roconnor?

**Comment by cstrahan** [comment]
> I wonder if it is safe to use the attribute name `override`?

I considered using `extend` instead (as I'll refer to it below), per my rationale above.

> Lastly, I'm not sure what `makeObject` brings to the table versus the existing `fix'` function. Can't we simply define a library function:

>     lib.overrideObject = ext: obj: fix' (lib.extends ext obj.__unfix__)

I'll have to give that some thought, and see if that addresses the other things I'd like:

* An intuitive interface (preferably similar to the `override` given by `callPackage`)
* A way to compose `override` and `extend` while preserving the "effects" of both.

Gotta run now, I'll give me thoughts in a little bit.

**Comment by roconnor** [comment]
> I don't know if it's so much confusing as it is unhelpful (can't imagine someone will want to do makeObject (self: /*...*/ self.override /*...*/)).

You can make this sort of nonsense impossible my moving the override attribute out of the fixpoint:

    let construct = rattrs: fix rattrs // {override = f: construct (extends f rattrs)}
    in construct

**Comment by domenkozar** [comment]
Could we also expose `kernelExtraConfig` as we do for grsecurity? Current method documented on wiki is very hard to type out without copy/paste :) 

https://nixos.org/wiki/How_to_tweak_Linux_kernel_config_options#Adding_options_to_the_mainstream_kernel

**Comment by cstrahan** [comment]
Sorry for letting this slip, had a big deadline at work, and I haven't been feeling well until recently. I can probably poke at this again in the next day or two.

File: lib/trivial.nix (hunk 5 above)
Comment (domenkozar) [inline lib/trivial.nix]: @cstrahan I'd really like to use this, but let's first simplify as @roconnor suggested :)

File: lib/trivial.nix [hunk 6]
```diff
@@ -53,6 +53,30 @@ rec {
...
+  #     nix-repl> obj = obj.override (self: super: { foo = "foo"; })
+  #
+  #     nix-repl> obj
+  #     { foo = "foo"; override = «lambda»; }
+  #
+  #     nix-repl> obj = obj.override (self: super: { foo = super.foo + " + "; bar = "bar"; foobar = self.foo + self.bar; })
+  #
+  #     nix-repl> obj
+  #     { bar = "bar"; foo = "foo + "; foobar = "foo + bar"; override = «lambda»; }
+  makeObject = rattrs:
```
Comment (domenkozar) [inline lib/trivial.nix]: There is no such things as Object in Nix.

A better name would be `mkOverridableAttrset`.

**Comment by domenkozar** [comment]
@cstrahan +1 for doing this. I'm [working on a package that depends on kernel outside nixpkgs](https://github.com/snabblab/snabblab-nixos/blob/master/pkgs/default.nix) and currently it's hard to extend `linuxPackages_*`. 

+1 for renaming `override` to `extend` and documenting it also in https://github.com/NixOS/nixpkgs/blob/master/doc/functions.xml

**Comment by cstrahan** [comment]
I rebased on top of the tip of master and made some of the suggested changes. I chose the word "extensible" (e.g. in `makeExtensible`) over "extendable" per https://stereochro.me/weblog/3246, despite the lack of symmetry between the words "extend" and "extensible":

> People get these works mixed up all the time, but, at least in technical circles, the difference is important.
>
> Let’s imagine you have a rod. The rod can be said to be extendible if it’s telescoping (like an antenna), or if it can be folded up and unfurled (like the support rods for a tent). The rod can be said to be extensible if you can put attachments onto the end to vary what the rod may be used for.
>
> And that’s why XML stands for ‘eXtensible Markup Language’ and why EPP stands for ‘Extensible Provisioning Protocol’. It’s not just some arbitrary choice of synonym.

@domenkozar How does this look now? If the changes look good, I can add docs in `functions.xml` when I have a spare moment.

**Comment by domenkozar** [comment]
I won't have time to look at this this week, but noone else does, I'll do my best next week.

**Comment by cstrahan** [comment]
@domenkozar Any chance you'll have some free time to take a look at this? :smile:

**Comment by domenkozar** [comment]
@cstrahan after 16.09 :)

**Comment by cstrahan** [comment]
Understandable, though it would be nice to have 16.09 :)

**Comment by domenkozar** [comment]
Agreed, but it's going to be released tomorrow so that ship is already gone. We can talk about backporting once it's stable enough in the master branch.

**Comment by Ericson2314** [comment]
Besides the merge conflict, this is ready now? I'm very interested in seeing this merged :)

**Comment by domenkozar** [comment]
Good job, @cstrahan rebase and merge :)

**Review by fpletz — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -11419,121 +11419,123 @@ in
      for a specific kernel.  This function can then be called for
      whatever kernel you're using. */
 
-  linuxPackagesFor = kernel: self: let callPackage = newScope self; in rec {
+  linuxPackagesFor = kernel: lib.makeExtensible (self: {
```
Comment (Ericson2314) [inline pkgs/top-level/all-packages.nix]: Come to think of it, we could just do `(self: with self; {` and cut down a lot of the diff.

**Comment by Ericson2314** [comment]
I did my changes and rebased in https://github.com/Ericson2314/nixpkgs/tree/kernel-patches

File: nixos/modules/system/boot/kernel.nix [hunk 8]
```diff
@@ -21,6 +23,11 @@ in
 
     boot.kernelPackages = mkOption {
       default = pkgs.linuxPackages;
+      apply = kernelPackages: kernelPackages.override (self: super: {
```
Comment (cstrahan) [inline nixos/modules/system/boot/kernel.nix]: Oh, yikes. Yes -- thanks for catching that. Didn't catch that after the renaming `override` to `extend`.

**Comment by cstrahan** [comment]
@Ericson2314 If you'd like to correct that broken reference to `override` (that should be `extend` instead) and push/merge you changes into master, I'm cool with that (actually, I'd appreciate it :wink:).

Alternatively, I have a deployment to tend to tonight, and I'll either have time much later tonight, or I can try to push this up tomorrow.

**Comment by Ericson2314** [comment]
@cstrahan I would do the honors, but I don't have push :).

**Comment by cstrahan** [comment]
Eh, I just went ahead and got it out of the way -- it only took a minute or two to rebase.

File: pkgs/top-level/all-packages.nix [hunk 9]
```diff
@@ -11419,121 +11419,123 @@ in
...
+    callPackage = newScope self;
+
     inherit kernel;
 
-    accelio = callPackage ../development/libraries/accelio { };
+    accelio = self.callPackage ../development/libraries/accelio { };
 
-    acpi_call = callPackage ../os-specific/linux/acpi-call {};
+    acpi_call = self.callPackage ../os-specific/linux/acpi-call {};
```
Comment (nbp) [inline pkgs/top-level/all-packages.nix]: I would prefer if this line could be read as `super.callPackage`, in order to avoid mistakes while using the overlays.

**Comment by edolstra** [comment]
It seems pretty ad hoc to have an option for applying patches to the kernel. Why not add an option for applying patches to Firefox, or any of the thousands of packages in Nixpkgs? This opens the door for a huge amount of option bloat.

**Comment by Ericson2314** [comment]
@edolstra I believe it is because Linux already was exposing parameters for patches, unlike most packages? This merely reexposes that parameter on the NixOS level.
